### PR TITLE
[#IOPID-1831, #IOPID-1840]  Expose getFIMSToken

### PR DESCRIPTION
**Pull request by gquadrati**
This PR depends on https://github.com/pagopa/io-infra/pull/976

#### List of Changes

- [[#IOPID-1831] add getByFIMSToken to redis service + tests](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/7993717d4fa1d5d1eaedb347eb5b6a1e2e27293a)
- [[#IOPID-1831] add fims.yaml + generation](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/1ae81ed2d4c34f50d9909707fb2e28cad48d505c)
- [[#IOPID-1831] add new SSOController + tests](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/b6b2349ed2b056ebf3c7a5925beaf40abb2f84e1)
- [[#IOPID-1831] add new bearer-FIMS-token-strategy](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/efd8317b103b11ac1d0a618db74322316929e9c2)
- [[#IOPID-1831] mount new route + update env-file](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/1090f78b835dfdaa6f6b5896556eae7cba756fd4)
- [[#IOPID-1831] fix after rebase + small refactor](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/2088a397b1d3d6333de78cfb002b58b22090dc37)
- [[#IOPID-1831] buyscout rule: fix lint on tests](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/2155be54d02653d9d84166e8df2673c526e7f601)

#### Motivation and Context

Expose getFIMSToken endpoint

⚠️ Another PR will follow addressing the necessary changes for FIMS+ integration

#### How Has This Been Tested?

Unit test, manual API call with io-mock

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts [hunk 1]
```diff
@@ -267,6 +269,110 @@ describe("RedisSessionStorage#getBySessionToken", () => {
   });
 });
 
+describe("RedisSessionStorage#getByFIMSToken", () => {
+  const mockedDependencies = {
+    redisClientSelector: mockRedisClientSelector,
+  };
+
+  test("should get the session with a valid token", async () => {
+    mockGet.mockImplementationOnce((_) => Promise.resolve(mockSessionToken));
```
Comment (arcogabbo) [inline apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts:278]: I would suggest this shortcut 😃 
```suggestion
    mockGet.mockResolvedValueOnce(mockSessionToken);
```

File: apps/session-manager/src/controllers/sso.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        err as Exclude<
+          RightOf<typeof response>,
+          IResponseSuccessJson<InitializedProfile>
+        >,
+    ),
+  );
+
+const toFIMSUser =
+  (deps: GetUserForFIMSDependencies) => (userProfile: InitializedProfile) =>
+    pipe(
```
Comment (BurnedMarshal) [inline apps/session-manager/src/controllers/sso.ts:88]:
```suggestion
const toFIMSUser: (
  deps: GetUserForFIMSDependencies,
) => RTE.ReaderTaskEither<
  InitializedProfile,
  IResponseErrorInternal,
  FIMSUser
> = (deps) => (userProfile) =>
  pipe(
```

File: apps/session-manager/src/services/redis-session-storage.ts [hunk 3]
```diff
@@ -25,6 +32,14 @@ import { SessionInfo } from "../generated/backend/SessionInfo";
...
 import { multipleErrorsFormatter } from "../utils/errors";
 import { RedisClientMode, RedisClientSelectorType } from "../types/redis";
+import {
+  bpdTokenPrefix,
+  fimsTokenPrefix,
+  myPortalTokenPrefix,
+  sessionNotFoundError,
+  walletKeyPrefix,
+  zendeskTokenPrefix,
+} from "../repositories/redis";
```
Comment (BurnedMarshal) [inline apps/session-manager/src/services/redis-session-storage.ts:42]: These should be already available in `RedisRepo`

File: apps/session-manager/src/controllers/__tests__/sso.spec.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  test("when the profile contains a validated email, then it returns a FIMS user with email", async () => {
+    await pipe(
+      mockedDependencies,
+      getUserForFIMS,
+      TE.map((response) => response.apply(res)),
+      TE.mapLeft((err) => expect(err).toBeFalsy()),
+    )();
+
+    expect(res.status).toHaveBeenCalledWith(200);
```
Comment (BurnedMarshal) [inline apps/session-manager/src/controllers/__tests__/sso.spec.ts:77]: This is good enough, but we can check the right path in a different way
```suggestion
    const result = await pipe(
      mockedDependencies,
      getUserForFIMS,
      TE.map((response) => response.apply(res)),
    )();

    expect(E.isLeft(result)).toBeFalsy();
    expect(res.status).toHaveBeenCalledWith(200);
```

File: apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts [hunk 5]
```diff
@@ -214,19 +220,15 @@ describe("RedisSessionStorage#getBySessionToken", () => {
   });
 
   test("should fail parse of user payload", async () => {
-    mockGet.mockImplementationOnce((_) => Promise.resolve("Invalid JSON"));
+    mockGet.mockImplementationOnce((_) => Promise.resolve(INVALID_JSON));
```
Comment (BurnedMarshal) [inline apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts:223]: a little bit more compact
```suggestion
    mockGet.mockResolvedValueOnce(INVALID_JSON);
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM, only some comments

File: apps/session-manager/src/controllers/__tests__/sso.spec.ts (hunk 4 above)
Comment (gquadrati) [reply apps/session-manager/src/controllers/__tests__/sso.spec.ts]: Done in [[#IOPID-1831] refactor based on comments](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/f7172071277aa811bf0b9d7c392ed059f7a54b29)

File: apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts (hunk 5 above)
Comment (gquadrati) [reply apps/session-manager/src/services/__tests__/redis-session-storage.spec.ts]: Done in [[#IOPID-1831] refactor based on comments](https://github.com/pagopa/io-auth-n-identity-domain/pull/21/commits/f7172071277aa811bf0b9d7c392ed059f7a54b29)

**Pull request merged**